staging/src/kubevirt.io/client-go/apis/snapshot/v1alpha1/types.go

```diff
@@ -163,11 +163,58 @@ type SourceSpec struct {
 	VirtualMachine *v1.VirtualMachine `json:"virtualMachine,omitempty"`
 }
 
+// EmbdededMetadata contains subset of standard object's metadata.
+type EmbdededMetadata struct {
+	// Name must be unique within a namespace. Is required when creating resources, although
+	// some resources may allow a client to request the generation of an appropriate name
+	// automatically. Name is primarily intended for creation idempotence and configuration
+	// definition.
+	// Cannot be updated.
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// Namespace defines the space within each name must be unique. An empty namespace is
+	// equivalent to the "default" namespace, but "default" is the canonical representation.
+	// Not all objects are required to be scoped to a namespace - the value of this field for
+	// those objects will be empty.
+	//
+	// Must be a DNS_LABEL.
+	// Cannot be updated.
+	// More info: http://kubernetes.io/docs/user-guide/namespaces
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Map of string keys and values that can be used to organize and categorize
+	// (scope and select) objects. May match selectors of replication controllers
+	// and services.
+	// More info: http://kubernetes.io/docs/user-guide/labels
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Annotations is an unstructured key value map stored with a resource that may be
+	// set by external tools to store and retrieve arbitrary metadata. They are not
+	// queryable and should be preserved when modifying objects.
+	// More info: http://kubernetes.io/docs/user-guide/annotations
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type PersistentVolumeClaim struct {
+	// Subset of standard object's metadata.
+	// +optional
+	EmbdededMetadata `json:"metadata,omitempty"`
```

Comment:
if the perserveUnknownFields is the primary issue here, do we still need this embdededMetadata field? why couldn't we do this?

```
        // +kubebuilder:pruning:PreserveUnknownFields
        // +nullable
        ObjectMeta metav1.ObjectMeta `json:"metadata,omitempty"`
```

also, should EmbdededMetadata be EmbeddedMetadata, or some other name? Not sure what embdeded is referring to.

Reply:
Sorry, it's a typo. It should be `EmbeddedMetadata` or `EmbeddedObjectMeta`. One benefit this solution has is that using our struct will render full schema with descriptions. It will also clearly state what we accept and validation will take a place. While in case of `+kubebuilder:pruning:PreserveUnknownFields` validation is turned off and schema looks like this:
```
metadata:
              nullable: true
              type: object
              x-kubernetes-preserve-unknown-fields: true
```
comparing to:
```
metadata:
                    description: Subset of standard object's metadata.
                    properties:
                      annotations:
                        additionalProperties:
                          type: string
                        description: 'Annotations is an unstructured key value map stored with a resource that may be set by external tools to store and retrieve arbitrary metadata. They are not queryable and should be preserved when modifying objects. More info: http://kubernetes.io/docs/user-guide/annotations'
                        type: object
                      labels:
                        additionalProperties:
                          type: string
                        description: 'Map of string keys and values that can be used to organize and categorize (scope and select) objects. May match selectors of replication controllers and services. More info: http://kubernetes.io/docs/user-guide/labels'
                        type: object
                      name:
                        description: Name must be unique within a namespace. Is required when creating resources, although some resources may allow a client to request the generation of an appropriate name automatically. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated.
                        type: string
                      namespace:
                        description: "Namespace defines the space within each name must be unique. An empty namespace is equivalent to the \"default\" namespace, but \"default\" is the canonical representation. Not all objects are required to be scoped to a namespace - the value of this field for those objects will be empty. \n Must be a DNS_LABEL. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/namespaces"
                        type: string
                    type: object
```
In our case the validation doesn't benefit us that much as we have our own validation webhooks(at least for some objects). But the clearly stated API can give a better user expirience since setting `GenerateName` would be meaningless but valid case(in `ObjectMeta`).

To answer your second question, we could do it but I see a benefit in this approach. WDYT?

Reply:
I don't feel strongly either way.  But the extra validation seems like a good thing

Reply:
the problem we're trying to solve here is to get controllerTools to generate schema for our persistentVolumeClaim struct. that's solved with or without this new objectMeta struct.

I don't have a clear enough understanding of the potential impact of creating a stripped down version of ObjectMeta will do in this case. In the interest of getting the change we need in (the new persistent volume claim struct) I'd suggest separating the two change into separate PRs.

If the change to object meta makes sense here, then it seems like it would make sense for our other objects that references ObjectMeta as well.

Reply:
@davidvossel that totally makes sense to me

Reply:
I mirrored the struct and add `+kubebuilder:pruning:PreserveUnknownFields` as we do in other places. I will have a look on all references of `ObjectMeta` and come up with resonable replacement for our usecases.